src/test/java/io/jenkins/plugins/analysis/core/util/EnvironmentResolverTest.java

```diff
@@ -0,0 +1,141 @@
+
+import java.util.Arrays;
+
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.*;
+
+import hudson.EnvVars;
+
+/**
+ * Unit test for {@link EnvironmentResolver}.
+ *
+ * @author Deniz Mardin
+ * @author Frank Christian Geyer
+ */
+class EnvironmentResolverTest {
+
+    private static final int NUMBER_OF_SIGNS = 1024;
+    private static final String KEY = "$KEY";
+    private static final String VALUE = "$VALUE";
+
+    /**
+     * Verifies that the output is empty for an empty input.
+     */
+    @Test
+    void shouldReturnUnmodifiedOutputForEmptyInput() {
+        final EnvironmentResolver environmentResolver = new EnvironmentResolver();
+        final String expanded = environmentResolver.expandEnvironmentVariables(new EnvVars(), "");
+
+        assertThat(expanded).isEmpty();
+    }
+
+    /**
+     * Verifies that the output is the same for a conventional input.
+     */
+    @Test
+    void shouldReturnSameOutputAsInput() {
+        final EnvironmentResolver environmentResolver = new EnvironmentResolver();
+        final String expanded = environmentResolver.expandEnvironmentVariables(new EnvVars(), "TestString");
+
+        assertThat(expanded).isEqualTo("TestString");
+    }
+
+    /**
+     * Verifies that the output is null for an input which is null.
+     */
+    @Test
+    void shouldReturnNullForNullAsInput() {
+        final EnvironmentResolver environmentResolver = new EnvironmentResolver();
+        final String expanded = environmentResolver.expandEnvironmentVariables(new EnvVars(), null);
+
+        assertThat(expanded).isNull();
+    }
+
+    /**
+     * Verifies that the output is the same as the input for an environment which is set to null.
+     */
+    @Test
+    void shouldReturnSameOutputAsInputForAnEnvironmentWhichIsNull() {
+        final EnvironmentResolver environmentResolver = new EnvironmentResolver();
+        final String expanded = environmentResolver.expandEnvironmentVariables(null, "TestStringForNull");
+
+        assertThat(expanded).isEqualTo("TestStringForNull");
+    }
+
+    /**
+     * Verifies that the output is null when both, the environment and the input are null.
+     */
+    @Test
+    void shouldReturnNullWhenBothParametersEnvironmentAndTheInputAreNull() {
+        final EnvironmentResolver environmentResolver = new EnvironmentResolver();
+        final String expanded = environmentResolver.expandEnvironmentVariables(null, null);
+
+        assertThat(expanded).isNull();
+    }
+
+    /**
+     * Verifies that the output is an adjusted value of a given special input, which is based on $-signs.
+     */
+    @Test
+    void shouldReturnAdjustedOutputForAGivenInputBasedOnDollarSigns() {
+        final EnvironmentResolver environmentResolver = new EnvironmentResolver();
+        final EnvVars envVars = new EnvVars();
+        envVars.put(KEY, VALUE);
+        final String previous = "$$Test$$Testing$TestString$$Testing$Test";
```

Comment:
Das war mit gar nicht bewusst, dass $$ durch $ ersetzt wird. (Das ist ja auch ein Feature der aufgerufenen Methode in Jenkins Core).

Wäre es in unserem Fall nicht leichter zu verstehen, wenn 

1. "Test" so bleib wie es ist 
2. "$KEY" durch "$VALUE" ersetzt wird. "$VALUE" aber nicht weiter
3. "$KEY" durch "$VALUE" ersetzt wird und dann "$VALUE" durch was weiteres

Ergibt drei Tests...